internal/resolution/variablesources/installed_package.go

```diff
@@ -1,71 +1,71 @@
-	allBundles  []*catalogmetadata.Bundle
-	successors  successorsFunc
-	bundleImage string
-}
-
-func (r *InstalledPackageVariableSource) GetVariables(_ context.Context) ([]deppy.Variable, error) {
-	// find corresponding bundle for the installed content
-	resultSet := catalogfilter.Filter(r.allBundles, catalogfilter.WithBundleImage(r.bundleImage))
-	if len(resultSet) == 0 {
-		return nil, r.notFoundError()
-	}
-
-	// TODO: fast follow - we should check whether we are already supporting the channel attribute in the operator spec.
-	//       if so, we should take the value from spec of the operator CR in the owner ref of the bundle deployment.
-	//       If that channel is set, we need to update the filter above to filter by channel as well.
-	sort.SliceStable(resultSet, func(i, j int) bool {
-		return catalogsort.ByVersion(resultSet[i], resultSet[j])
-	})
-	installedBundle := resultSet[0]
-
-	upgradeEdges, err := r.successors(r.allBundles, installedBundle)
-	if err != nil {
-		return nil, err
-	}
-
-	// you can always upgrade to yourself, i.e. not upgrade
-	upgradeEdges = append(upgradeEdges, installedBundle)
-	return []deppy.Variable{
-		variables.NewInstalledPackageVariable(installedBundle.Package, upgradeEdges),
-	}, nil
-}
-
-func (r *InstalledPackageVariableSource) notFoundError() error {
-	return fmt.Errorf("bundleImage %q not found", r.bundleImage)
-}
-
-func NewInstalledPackageVariableSource(allBundles []*catalogmetadata.Bundle, bundleImage string) (*InstalledPackageVariableSource, error) {
-	successors := legacySemanticsSuccessors
+// MakeInstalledPackageVariables returns variables representing packages
+// already installed in the system.
+// Meaning that each BundleDeployment managed by operator-controller
+// has own variable.
+func MakeInstalledPackageVariables(
+	allBundles []*catalogmetadata.Bundle,
+	bundleDeployments []rukpakv1alpha1.BundleDeployment,
+) ([]*olmvariables.InstalledPackageVariable, error) {
+	var successors successorsFunc = legacySemanticsSuccessors
 	if features.OperatorControllerFeatureGate.Enabled(features.ForceSemverUpgradeConstraints) {
 		successors = semverSuccessors
 	}
 
-	return &InstalledPackageVariableSource{
-		allBundles:  allBundles,
-		bundleImage: bundleImage,
-		successors:  successors,
-	}, nil
+	result := make([]*olmvariables.InstalledPackageVariable, 0, len(bundleDeployments))
+	processed := sets.Set[string]{}
+	for _, bundleDeployment := range bundleDeployments {
+		sourceImage := bundleDeployment.Spec.Template.Spec.Source.Image
+		if sourceImage == nil || sourceImage.Ref == "" {
+			continue
+		}
+
+		if processed.Has(sourceImage.Ref) {
+			continue
+		}
+		processed.Insert(sourceImage.Ref)
+
+		bundleImage := sourceImage.Ref
+
+		// find corresponding bundle for the installed content
+		resultSet := catalogfilter.Filter(allBundles, catalogfilter.WithBundleImage(bundleImage))
+		if len(resultSet) == 0 {
+			return nil, fmt.Errorf("bundleImage %q not found", bundleImage)
```

Comment:
When this error surfaces, the reader won't know what `bundleImage` is or necessarily where it's coming from. Would it be possible to revise the text to try to minimize confusion?

Reply:
This one is a bit tricky becuase this is a situaion when we have something installed on the cluster, but do not know what it is.

I imagine we can end up here with steps like this:

* User creates a catalog
* Installs a pacakge
* Removes the catalog

At this point we are unable to determine what bundle was installed.

For #449 I'll (likely) be modifying this function so that we will have a relevant `Operator` here. After this change we will probably be able to use `Operator`'s spec to add more context (e.g. package name).

But there are still things related to #449 I'm not sure about. For example, today we take all `BundleDeployment`s into account (even ones created directly). Doing the above will likely mean that we will only care about `BundleDeployment`s which we manage (ones owned by an `Operator`). So this is a bit in the air at the moment. Hopefully I'll get to it next week. Relevant discussions:
* https://github.com/operator-framework/operator-controller/pull/460#discussion_r1376252897
* https://github.com/operator-framework/operator-controller/issues/443

Reply:
Even though we don't have bundle or package information I think we have enough information on hand to provide a more clear error message that gives someone enough information to take some kind of remediation action if necessary. Please correct me if I'm wrong, but my assumption here is that this results in a failed resolution and prevents an Operator from successfully installing.

Maybe something like:
```suggestion
			return nil, fmt.Errorf("bundle with image %q not found in set of installable content but is installed via BundleDeployment %q", bundleImage, bundleDeployment.Name)
```

Reply:
I just remembered that there was an idea to add labels to `BundleDeployment` to contain package name, bundle name, catalog name. It will help a bit, but we will still have an issue with BDs created derectly.

@everettraven yeah, it looks like having a BD's name in here will be helpful. I'll take a look into your suggestion on Monday.

---

I'll probably create a separate PR for this because most likely some tests will have to be updated as well.

Please let's keep this thread open until I come back to this on Monday.

Reply:
Created a separate PR for this: https://github.com/operator-framework/operator-controller/pull/514

It will need a rebase once we merge this one.